pkg/ocgormv2/callbacks.go

```diff
@@ -0,0 +1,231 @@
+		db.Callback().Query().After("gorm:query").Register("instrumentation:after_query", c.afterQuery),
+		db.Callback().Query().Before("gorm:row_query").Register("instrumentation:before_row_query", c.beforeRowQuery),
+		db.Callback().Query().After("gorm:row_query").Register("instrumentation:after_row_query", c.afterRowQuery),
+		db.Callback().Update().Before("gorm:update").Register("instrumentation:before_update", c.beforeUpdate),
+		db.Callback().Update().After("gorm:update").Register("instrumentation:after_update", c.afterUpdate),
+		db.Callback().Delete().Before("gorm:delete").Register("instrumentation:before_delete", c.beforeDelete),
+		db.Callback().Delete().After("gorm:delete").Register("instrumentation:after_delete", c.afterDelete))
+}
+
+func (c *callbacks) before(db *gorm.DB, operation string) {
+	ctx := db.Statement.Context
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	ctx = c.startTrace(ctx, db, operation)
+	ctx = c.startStats(ctx, db, operation)
+
+	db.Statement.Context = ctx
+}
+
+func (c *callbacks) after(db *gorm.DB) {
+	c.endTrace(db)
+	c.endStats(db)
+}
+
+func (c *callbacks) startTrace(ctx context.Context, db *gorm.DB, operation string) context.Context {
+	// Context is missing, but we allow root spans to be created
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	parentSpan := trace.FromContext(ctx)
+	if parentSpan == nil && !c.allowRoot {
+		return ctx
+	}
+
+	var span *trace.Span
+
+	if parentSpan == nil {
+		ctx, span = trace.StartSpan(
+			context.Background(),
+			fmt.Sprintf("gorm:%s", operation),
+			trace.WithSpanKind(trace.SpanKindClient),
+			trace.WithSampler(c.startOptions.Sampler),
+		)
+	} else {
+		ctx, span = trace.StartSpan(ctx, fmt.Sprintf("gorm:%s", operation))
+	}
+
+	attributes := append(
+		c.defaultAttributes,
+		trace.StringAttribute(ocgorm.TableAttribute, db.Statement.Table),
+	)
+
+	if c.query {
+		attributes = append(attributes, trace.StringAttribute(ocgorm.ResourceNameAttribute, db.Statement.SQL.String()))
+	}
+
+	span.AddAttributes(attributes...)
+
+	return ctx
+}
+
+func (c *callbacks) endTrace(db *gorm.DB) {
+	span := trace.FromContext(db.Statement.Context)
+
+	// Add query to the span if requested
+	if c.query {
+		span.AddAttributes(trace.StringAttribute(ocgorm.ResourceNameAttribute, db.Statement.SQL.String()))
+	}
+
+	var status trace.Status
+
+	if db.Error != nil {
+		if errors.Is(db.Error, gorm.ErrRecordNotFound) {
+			status.Code = trace.StatusCodeNotFound
+		} else {
+			status.Code = trace.StatusCodeUnknown
+		}
+
+		status.Message = db.Error.Error()
+	}
+
+	span.SetStatus(status)
+
+	span.End()
+}
+
+var (
+	queryStartPropagator, _ = tag.NewKey("sql.query_start")
+)
+
+func (c *callbacks) startStats(ctx context.Context, db *gorm.DB, operation string) context.Context {
+	ctx, _ = tag.New(ctx,
+		tag.Upsert(ocgorm.Operation, operation),
+		tag.Upsert(ocgorm.Table, db.Statement.Table),
+		tag.Upsert(queryStartPropagator, time.Now().UTC().Format(time.RFC3339Nano)),
```

Comment:
Is this a metric, and if so, are we tagging the metric with `time.Now()`?

Or is this a trace tag?

Reply:
Yes, It is used to record the start time of a database operation.  Later, this start time is retrieved from the context, and If the start time exists, it is parsed back into a `time.Time` value. The elapsed time of the operation is then calculated by subtracting this start time from the current time. This elapsed time is recorded as a metric, specifically as the latency of the operation in milliseconds. 

This code is not new,  it already existed in a previous version of the library utilized within the cloud-sdk. This addition is only to support for Gorm v2.

Reply:
Thanks! I think this originally spooked me because it looked like it was part of a _metric_, but if I'm reading this right, it's part of a _trace_ span.